### PR TITLE
StreamAllocator tweaks

### PR DESCRIPTION
**Pull request by boks1971**
- Fix issue with video corruption - was not setting slice length properly when munging VP8
- Forward padding only packets after a gap in sequence number
- Feature flag to control streaming when switching from higher layer -> lower layer

### CODE REVIEW

File: pkg/sfu/downtrack.go [hunk 1]
```diff
@@ -873,8 +875,8 @@ func (d *DownTrack) translateVP8PacketTo(pkt *rtp.Packet, incomingVP8 *buffer.VP
 		copy(dstPayload, srcPayload)
 	}
 
-	hdr := buf[:translatedVP8.HeaderSize]
-	return translatedVP8.MarshalTo(hdr)
+	err := translatedVP8.MarshalTo(buf[:translatedVP8.HeaderSize])
+	return buf, err
```
Comment (boks1971) [inline pkg/sfu/downtrack.go:879]: This is the main fix for video corruption. Was not returning buf.

File: pkg/sfu/forwarder.go [hunk 2]
```diff
@@ -692,7 +792,7 @@ func (f *Forwarder) getTranslationParamsVideo(extPkt *buffer.ExtPacket, layer in
 		return tp, nil
 	}
 
-	if f.targetLayers.spatial < f.currentLayers.spatial && f.targetLayers.spatial < f.maxLayers.spatial {
+	if FlagPauseOnDowngrade && f.targetLayers.spatial < f.currentLayers.spatial && f.targetLayers.spatial < f.maxLayers.spatial {
```
Comment (boks1971) [inline pkg/sfu/forwarder.go:795]: Making this a feature flag, i. e. pausing higher layer when switching to lower layer - defaulting to false for now to enable smoother transition

File: pkg/sfu/forwarder.go [hunk 3]
```diff
@@ -764,7 +864,7 @@ func (f *Forwarder) getTranslationParamsVideo(extPkt *buffer.ExtPacket, layer in
 		return tp, err
 	}
 
-	if f.vp8Munger == nil {
+	if f.vp8Munger == nil || len(extPkt.Packet.Payload) == 0 {
```
Comment (boks1971) [inline pkg/sfu/forwarder.go:867]: Do not parse as VP8 padding only packets and forward them if it gets here. If padding only packets get here, it is following a gap in sequence number. As we do not know what temporal layer would have been in the gap, just forward unconditionally. We may be able to drop padding only packets even after a gap. Need some more testing/thinking. But, forwarding for now.

**Review by ashellunts — APPROVED** [review]

**Comment by boks1971** [comment]
Thank you @ashellunts . @cnderrauber @davidzhao Going to merge this. But, as usual please review and comment. Will take care of things in a subsequent PR.

**Comment by boks1971** [comment]
doh, had a local change with some commented out code deleted. Forgot to push. Will take care of it next round.

**Pull request merged**